### PR TITLE
PORTALS-2988 - Fix cache invalidation issues

### PR DESCRIPTION
**Pull request by nickgros**
- Fix mismatch issues with entity query keys in KeyFactory
- Add tests to KeyFactory to help prevent regressions
- Rename QueryClientUtils to QueryFilterUtils
- Add to QueryFilterUtils `getAllActionsRequiredQueryFilters`
- Update query invalidation for actions required to use `getAllActionsRequiredQueryFilters`, which now includes table queries that include the relevant bit in the partMask, fixing PORTALS-2988
- Fix memoization issue in useGetQueryResultBundleWithAsyncStatus, which could trigger performance problems in components that use table query data.

### CODE REVIEW

File: packages/synapse-react-client/src/synapse-queries/KeyFactory.test.ts [hunk 1]
```diff
@@ -0,0 +1,596 @@
...
+  BUNDLE_MASK_QUERY_RESULTS,
+} from '../utils/SynapseConstants'
+
+describe('KeyFactory tests', () => {
+  let keyFactory = new KeyFactory(MOCK_ACCESS_TOKEN)
+  beforeEach(() => {
+    keyFactory = new KeyFactory(MOCK_ACCESS_TOKEN)
+  })
+
+  describe('Entity QueryKeys', () => {
```
Comment (nickgros) [inline packages/synapse-react-client/src/synapse-queries/KeyFactory.test.ts:20]: Added many tests to help prevent regressions. Not yet comprehensive

File: packages/synapse-react-client/src/synapse-queries/KeyFactory.ts [hunk 2]
```diff
@@ -122,7 +122,7 @@ const entityQueryKeyObjects = {
     infinite: boolean,
   ) => [
     entityQueryKeyObjects.entity(queryBundleRequest.entityId),
-    'tableQueryResultWithAsyncStatus',
+    { type: 'tableQueryResult', withAsyncStatus: true },
```
Comment (nickgros) [inline packages/synapse-react-client/src/synapse-queries/KeyFactory.ts:125]: Update the `tableQueryResult` query keys to all match so they can all be invalidated at once, when necessary

File: packages/synapse-react-client/src/synapse-queries/QueryClientUtils.ts
Comment (nickgros) [inline packages/synapse-react-client/src/synapse-queries/QueryClientUtils.ts:1]: Renamed

File: packages/synapse-react-client/src/synapse-queries/QueryFilterUtils.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  id: string,
+) {
+  // TODO: Figure out how to handle cases where the change can affect other entities.
+  // Some examples:
+  //  - deleting an entity should invalidate the parent's children
+  //  - moving an entity should invalidate the old and new parent's children
+  return queryClient.invalidateQueries({
+    queryKey: keyFactory.getEntityQueryKey(id),
+  })
+}
```
Comment (nickgros) [inline packages/synapse-react-client/src/synapse-queries/QueryFilterUtils.ts:25]: This was just moved from the former file name `QueryClientUtils`, with no changes

File: packages/synapse-react-client/src/synapse-queries/QueryFilterUtils.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+/**
+ * Invalidate all queries that retrieve the actions required to gain download access
+ * @param queryClient
+ * @param keyFactory
+ */
+export function getAllActionsRequiredQueryFilters(
```
Comment (nickgros) [inline packages/synapse-react-client/src/synapse-queries/QueryFilterUtils.ts:32]: New

File: packages/synapse-react-client/src/synapse-queries/QueryFilterUtils.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+              queryKeyItem.tableQueryBundleRequest.partMask &
+                BUNDLE_MASK_ACTIONS_REQUIRED,
+            )
+          ) {
+            return true
+          }
+        }
+        return false
+      },
+    },
```
Comment (nickgros) [inline packages/synapse-react-client/src/synapse-queries/QueryFilterUtils.ts:68]: This was missing, which caused PORTALS-2988 -- invalidate query bundle requests when the action results change

File: packages/synapse-react-client/src/synapse-queries/auth/useTwoFactorEnrollment.ts [hunk 6]
```diff
@@ -52,13 +53,9 @@ export function useFinishTwoFactorEnrollment(
...
-        // including for entities and the download list
-        queryClient.invalidateQueries({
-          queryKey: keyFactory.getAllEntityActionsRequiredQueryKey(),
-        }),
-        queryClient.invalidateQueries({
-          queryKey: keyFactory.getDownloadListBaseQueryKey(),
-        }),
+        ...getAllActionsRequiredQueryFilters(keyFactory).map(filter =>
+          queryClient.invalidateQueries(filter),
+        ),
```
Comment (nickgros) [inline packages/synapse-react-client/src/synapse-queries/auth/useTwoFactorEnrollment.ts:58]: Use the new utility where actions required may change

File: packages/synapse-react-client/src/synapse-queries/KeyFactory.test.ts (hunk 1 above)
Comment (jay-hodgson) [reply packages/synapse-react-client/src/synapse-queries/KeyFactory.test.ts:20]: Nice "start"!

**Review by jay-hodgson — APPROVED** [review]

**Pull request merged**